stdlib/dynarray.mli

```diff
@@ -0,0 +1,615 @@
+    or equal to the length of the dynamic array. We call {b capacity}
+    the length of the backing array.
+
+    The capacity of a dynamic array is relevant in advanced scenarios,
+    when reasoning about the performance of dynamic array programs:
+    {ul
+    {- The memory usage of a dynamic array is proportional to its capacity,
+       rather than its length.}
+    {- When then is no empty space left at the end of the backing array.
+       adding elements requires allocating a new, larger backing array.}}
+
+    The implementation uses a standard exponential reallocation
+    strategy which guarantees amortized constant-time operation: the
+    total capacity of all backing arrays allocated over the lifetime
+    of a dynamic array is at worst proportional to the total number of
+    elements added.
+
+    In other words, users need not care about capacity and reallocations,
+    and they will get reasonable behavior by default. However, in some
+    performance-sensitive scenarios the functions below can help control
+    memory usage or guarantee an optimal number of reallocations.
+*)
+
+val capacity : 'a t -> int
+(** [capacity a] is the length of [a]'s backing array. *)
+
+val ensure_capacity : 'a t -> int -> unit
+(** [ensure_capacity a n] makes sure that the capacity of [a]
+    is at least [n].
+
+    @raise Invalid_argument if the requested capacity is
+      outside the range [0 .. Sys.max_array_length].
+
+    An example would be to implement {!of_array} (without using {!init} directly):
+    {[
+    let of_array arr =
+      let a = Dynarray.create () in
+      Dynarray.ensure_capacity a (Array.length arr);
+      Array.iter (fun v -> add_last a v) arr
+    ]}
+
+    Using [ensure_capacity] guarantees that at most one reallocation
+    will take place, instead of possibly several.
+
+    Without this [ensure_capacity] hint, the number of resizes would
+    be logarithmic in the length of [arr], creating a constant-factor
+    slowdown noticeable when [arr] is large.
+*)
+
+val ensure_extra_capacity : 'a t -> int -> unit
+(** [ensure_extra_capacity a n] is [ensure_capacity a (length a + n)],
+    it makes sure that [a] has room for [n] extra items.
+
+    A use case would be to implement {!append_array}:
+    {[
+    let append_array a arr =
+      ensure_extra_capacity a (Array.length arr);
+      Array.iter (fun v -> add_last a v) arr
+    ]}
+*)
+
+val fit_capacity : 'a t -> unit
+(** [fit_capacity a] shrinks the backing array so that its capacity is
+    exactly [length a], with no additional empty space at the
+    end. This can be useful to make sure there is no memory wasted on
+    a long-lived array.
+
+    Note that calling [fit_capacity] breaks the amortized complexity
+    guarantees provided by the default reallocation strategy. Calling
+    it repeatedly on an array may have quadratic complexity, both in
+    time and in total number of allocations.
```

Comment:
"Size" is ambiguous: it could be understood as "memory size, as observed after the reuse caused by the GC". "Number of allocations" is more precise, but less interesting (it is bounded by "time").

Reply:
I reformulated and also avoided the word "shrink" as suggested.

> "Number of allocations" is more precise, but less interesting (it is bounded by "time").

In fact, in "quadratic in time and total number of words allocated", it is "time" that is less interesting / redundant. If I only tell you that something takes quadratic time, you don't know that it also produces unnecessary GC churn. I still decided to keep both for clarity.